examples/custom_distance/example_custom_distance.cpp

```diff
@@ -0,0 +1,142 @@
+  return Kokkos::abs(custom_point[0] - point[0]) * 2 +
+         Kokkos::abs(custom_point[1] - point[1]) / 2;
+}
+
+// Provide the distance function between the bounding volume geometry and
+// the custom data type
+KOKKOS_FUNCTION auto distance(CustomPoint const &point, Box const &box)
+{
+  static_assert(DIM == 2);
+  Point projected_point;
+  for (int d = 0; d < DIM; ++d)
+    projected_point[d] =
+        Kokkos::clamp(point[d], box.minCorner()[d], box.maxCorner()[d]);
+  return distance(point, projected_point);
+}
+
+// Allow simple wrapping of regular points into the custom data
+template <typename Points>
+struct CustomPoints
+{
+  Points points;
+};
+
+template <typename Points>
+struct ArborX::AccessTraits<CustomPoints<Points>, ArborX::PredicatesTag>
+{
+  using Self = CustomPoints<Points>;
+  using memory_space = typename Points::memory_space;
+
+  // Search for the two nearest neighbors
+  static constexpr int k = 2;
+
+  static KOKKOS_FUNCTION auto size(Self const &x) { return x.points.size(); }
+  static KOKKOS_FUNCTION decltype(auto) get(Self const &x, int i)
+  {
+    return ArborX::nearest(Kokkos::bit_cast<CustomPoint>(x.points(i)), k);
+  }
+};
+
+// Callback to store the resulting distances
+struct DistanceCallback
+{
+  template <typename Query, typename Value, typename Output>
+  KOKKOS_FUNCTION void operator()(Query const &query, Value const &value,
+                                  Output const &out) const
+  {
+    out(distance(ArborX::getGeometry(query), value));
+  }
+};
+
+int main(int argc, char *argv[])
+{
+  Kokkos::ScopeGuard guard(argc, argv);
+
+  using ExecutionSpace = Kokkos::DefaultExecutionSpace;
+  using MemorySpace = ExecutionSpace::memory_space;
+
+  ExecutionSpace space;
+
+  // In L2 metric, two bottom points are the closest.
+  // In custom metric, left two points are the closest.
+  //
+  // x
+  //
+  // o
+  // x x
+  Kokkos::View<Point *, MemorySpace> points("Example::points", 3);
+  auto points_host = Kokkos::create_mirror_view(points);
+  points_host[0] = {0, 0};
+  points_host[1] = {1, 0};
+  points_host[2] = {0, 3};
+  Kokkos::deep_copy(points, points_host);
+
+  Kokkos::View<Point *, MemorySpace> query_points("Example::query_points", 1);
+  auto query_points_host = Kokkos::create_mirror_view(query_points);
+  query_points_host[0] = {0, 1};
+  Kokkos::deep_copy(query_points, query_points_host);
+
+  ArborX::BVH<MemorySpace, Point> bvh(space, points);
+
+  Kokkos::View<float *, MemorySpace> distances("Example::distances", 0);
+  Kokkos::View<int *, MemorySpace> offsets("Example::offsets", 0);
+  bvh.query(space, CustomPoints<decltype(query_points)>{query_points},
+            DistanceCallback{}, distances, offsets);
+
+  auto offsets_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, offsets);
+  auto distances_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, distances);
+
+  // Expected output:
+  // offsets: 0 2
+  // distances: 0.5 1
```

Comment:
Why did you elect to get distances out instead of indices?
I suspect indices would make it more obvious that the distance metric was changed.
I find the use of both a custom geometry with specialized distance computation mixed with a callback that recomputes the distance potentially confusing.

Reply:
I chose distances so that a user can see that the computed distances are indeed the custom ones. Computing just the found indices seems insufficient to me, as it still does not confirm that we can get the correct distance in the callback.